contrib/kubespray/script/openpai-generator.py

```diff
@@ -173,32 +268,40 @@ def wait_amd_device_plugin_ready(total_time=3600):
             sys.exit(1)
 
 
-def hived_config_prepare(worker_dict, node_resource_dict):
+def hived_config_prepare(worker_dict, node_resource_dict, pai_daemon_resource_dict):
     hived_config = dict()
     hived_config["nodelist"] = []
 
     min_mem = 100000000
     min_gpu = 100000000
     min_cpu = 100000000
 
-    for key in node_resource_dict:
+
+    node_resource_free = node_resource_dict["free"]
+    node_resource_allocatable = node_resource_dict["allocatable"]
+    for key in node_resource_dict["free"]:
         if key not in worker_dict:
             continue
-        if node_resource_dict[key]["gpu-resource"] == 0:
+        if node_resource_free[key]["gpu-resource"] == 0:
             logger.error("Allocatable GPU number in {0} is 0, current quick start script does not allow.".format(key))
             logger.error("Please remove {0} from your workerlist, or check if the device plugin is running healthy on the node.".format(key))
             sys.exit(1)
-        min_cpu = min(min_cpu, node_resource_dict[key]["cpu-resource"])
-        min_mem = min(min_mem, node_resource_dict[key]["mem-resource"])
-        min_gpu = min(min_gpu, node_resource_dict[key]["gpu-resource"])
+        reserved_cpu = min(node_resource_allocatable[key]["cpu-resource"] * Decimal(PAI_RESERVE_RESOURCE_PERCENTAGE), PAI_MAX_RESERVE_CPU_PER_NODE)
+        reserved_mem = min(node_resource_allocatable[key]["mem-resource"] * Decimal(PAI_RESERVE_RESOURCE_PERCENTAGE), PAI_MAX_RESERVE_MEMORY_PER_NODE)
+        min_cpu = min(min_cpu, node_resource_free[key]["cpu-resource"] - pai_daemon_resource_dict["cpu-resource"] - reserved_cpu)
+        min_mem = min(min_mem, node_resource_free[key]["mem-resource"] - pai_daemon_resource_dict["mem-resource"] - reserved_mem)
+        min_gpu = min(min_gpu, node_resource_free[key]["gpu-resource"])
+        if min_cpu <= 0 or min_mem <= 0:
+            logger.error("The node resource is not satisfy minmal requests, we need at least cpu: %s, mem: %sMB", min_cpu, min_mem)
+            sys.exit(1)
         hived_config["nodelist"].append(key)
     if not hived_config["nodelist"]:
         logger.error("No worker node is detected.")
         sys.exit(1)
 
     hived_config["min-gpu"] = min_gpu
-    hived_config["unit-cpu"] = int( min_cpu / min_gpu )
-    hived_config["unit-mem"] = int( min_mem / min_gpu )
+    hived_config["unit-cpu"] = math.ceil( min_cpu / min_gpu )
+    hived_config["unit-mem"] = math.ceil( min_mem / min_gpu )
```

Comment:
We should use floor here, so user request are always within free.

Reply:
See step1 of https://github.com/microsoft/hivedscheduler/blob/master/doc/user-manual.md#config-quickstart

Reply:
my wrong, fixed